### PR TITLE
feat(ci): use ci for automated build, deploy and more 

### PR DESCRIPTION
**Pull request by pwnwriter**
Hello @0atman ,

First off, I wanna thank you for all the videos you've made, They are super helpful. I never thought I'd be able to make a pr to someone from whom I've learned this many things, but anyways lets goo.

I'll explain what I did.

- First it'll be using automated ci to build and push to `gh-pages` branch on `push` event via `main` branch. 
- I use nix on `apple silicon` so I had to use `flakes` for my `devenv`, again now ci uses `nix develop` for handling dependencies and so.
- I tried* making the entry a bit `error` handling. 
- The automated build and push needs, read and write permissions. You'll want to allow that from your, `repo settings >> Actions >> General >> Workflow permissions` 
- Fixed the YouTube video link ? probably ? 

POC: 
- [Automated deployment ci in action](https://github.com/pwnwriter/HumanMadeMarkcom/actions/runs/10017977899/job/27692803510)

- [deployed website preview](https://pwnwriter.github.io/HumanMadeMark-com)

If you find anything confusing or something needs to be changed please drop that. I'll be more than happy to follow up. 🤗

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+      - name: Checkout code
+        uses: actions/checkout@v3
+       
+      - name: Install Nix package manager
+        uses: DeterminateSystems/nix-installer-action@main
+
+      - name: Build the website
+        run: |
+          nix develop --command cargo run --verbose
+          nix develop --command cargo build --verbose
```
Comment (0atman) [inline .github/workflows/build.yml]: run implies build, I think line is redundant?

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+      - name: Deploy build to gh-pages branch
+        uses: crazy-max/ghaction-github-pages@v4
+        with:
+          target_branch: gh-pages
+          build_dir: docs
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Comment (0atman) [inline .github/workflows/build.yml]: so I should create a token with a limited scope to use this auto-builder, got it. I'll audit the action code before doing this.

File: README.md [hunk 3]
```diff
@@ -12,4 +12,4 @@ I've made the following video to try to make my reasons for making this clear, b
 I don't hate AIs,
 I love humans!
 
-Check out my video here: https://www.youtube.com/watch?v=y3wVvDbJHrU
+Check out my video here: https://youtu.be/qjPAWy_2FL0
```
Comment (0atman) [inline README.md]: ha! I fixed this in main too, but thanks for the driveby :sweat_smile:

File: flake.nix [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+              ] ++ (if builtins.elem system [ "x86_64-darwin" "aarch64-darwin" ] then [ pkgs.darwin.apple_sdk.frameworks.Security ] else []);
+            in
+            {
+              default = pkgs.mkShell {
+                buildInputs = buildInputs;
+
+                packages = with pkgs; [
+                   cargo
+                   rustc
+                   rustfmt
```
Comment (0atman) [inline flake.nix]: probably don't need rustfmt, right? and I imagine `cargo` implies `rustc` ?

File: src/main.rs [hunk 5]
```diff
@@ -271,8 +271,21 @@ fn build(pages: Router) -> Result<(), Report> {
...
+    }
+
+    match build(vec![("docs/index.html", index)]) {
+        Ok(_) => println!("Built site successfully."),
+        Err(e) => {
+            eprintln!("Error building site: {:?}", e);
+            return Err(e.into());
+        }
+    }
+
```
Comment (0atman) [inline src/main.rs]: nice improvement in error handling

**Review by 0atman — CHANGES_REQUESTED** [review]
Nearly everything here's terrific, thank you, but if it's not too much bother, please put rust code tweaks in a separate PR. (I disagree with some choices and would like to talk about them without holding up the auto-building nix action. Thanks again!

File: src/main.rs (hunk 5 above)
Comment (0atman) [reply src/main.rs]: though let's add this in another PR

File: README.md (hunk 3 above)
Comment (0atman) [reply README.md]: (code changes in another pr plz)

File: .github/workflows/build.yml (hunk 1 above)
Comment (pwnwriter) [reply .github/workflows/build.yml]: The `cargo build` doesn't actually build the site but the `cargo run` does. I've tested on my end. I recommend checking once by removing all the files inside `docs` directory to reproduce the issue. I thought it's a known issue.

File: .github/workflows/build.yml (hunk 2 above)
Comment (pwnwriter) [reply .github/workflows/build.yml]: Yo do not necessarily have to use GitHub auth token if you allow `read and write` perms?

File: README.md (hunk 3 above)
Comment (pwnwriter) [reply README.md]: sure thing.

File: flake.nix (hunk 4 above)
Comment (pwnwriter) [reply flake.nix]: While installing `rust` with `rustup`, it does install them as component but here's I'm installing them a standalone binaries,  so yes we need that here as well.

File: src/main.rs (hunk 5 above)
Comment (pwnwriter) [reply src/main.rs]: Okay

**Comment by pwnwriter** [comment]
Maybe we need to fix this as well? 

Screenshot 2024-07-22 at 4 28 06 PM

**Comment by 0atman** [comment]
I thought we might be talking at cross-purposes: This rust project is a static site generator. build builds the generator, whereas run, runs the static site generator, updating the html files, which are then served statically by gh-pages.

This is why I said you don't have to run `build` after `run` - just run it! :grin:

**Comment by pwnwriter** [comment]
The `cargo build --verbose` command also checks for issues with the code itself(building?), given that we don't have any test cases. If you prefer, I can remove this build.

**Comment by 0atman** [comment]
AFAIK `cargo run` on changed code will run `build` and all the checking there. Am I wrong? 

In the future we could integrate `clippy` to ensure very strict standards, but that's outside the scope of this pr I think. (my code might fail some lints lol)

**Comment by pwnwriter** [comment]
gotcha, wil make a separate pr for the same.

**Comment by 0atman** [comment]
super, thank you again! :smiley:

**Pull request closed without merging**